### PR TITLE
[WIP] Containerize broadway-api & deploy script

### PR DESCRIPTION
**Pull request by zhengyao-lin**
WIP:
- [x] Containerize broadway-api
- [x] Containerize broadway-grader
- [x] Setup script

Dockerized api & worker are pushed to Docker Hub:
`illinois241/broadway-api`, `illinois241/broadway-grader`

TODO list for a master node:
1. Install docker
2. Create database dir `$DB_PATH`
3. Run `mongo:4.1-xenial` with `--net=host -v <host db path>:/data/db mongo:4.1-xenial mongod --bind_ip=127.0.0.1 --port=<mongo port>`
4. Run broadway-api container with `--net=host api --db-uri=mongodb://localhost:<mongo port>`
5. Get the cluster token

TODO list for a worker node:
1. Install docker
2. Run dockerd with `-H tcp://localhost:<docker_port>`
3. Run broadway-grader container with `--net=host -v /tmp:/tmp -e DOCKER_HOST=tcp://localhost:<docker_port> grader <master host> <master port> <cluster token>`
4. Need to disable AppArmor(otherwise dockerd would fail to kill containers)

Update: deploy script https://github.com/rod-lin/broadway-man

### CODE REVIEW

**Comment by ayushr2** [comment]
How are we making the `broadway-grader` not run nested containers? If we are using the Docker API on the machines running the `broadway-grader` then how are we communicating results between the containers and the grader? And most importantly, how are we passing files between subsequent containers of a pipeline?

**Comment by rod-lin** [comment]
Oh right the grader has to bind /job to the container... Let me check.

(btw it's a WIP, I was just writing my draft. plz don't review..

File: src/constants/keys.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 # these let us specify the stage specs
 IMAGE = "image"
 ENV = "env"
-ENTRY_POINT = "entry_point"
+ENTRY_POINT = "entrypoint"
```
Comment (ayushr2) [inline src/constants/keys.py]: When had you branched off? This seems to be a change already merged into the `master` branch to make it compliant with chainlink.

https://github.com/illinois-cs241/broadway-api/blob/master/src/constants/keys.py#L34

Comment (ayushr2) [reply src/constants/keys.py]: Looks like some changes are leading to conflicts with master.

Comment (rod-lin) [reply src/constants/keys.py]: my bad. should be fixed now

**Comment by ayushr2** [comment]
Also, note that the containers have to be run with a strict timeout policy. You should be able to kill containers after they timeout.

Clients should be able to specify these options for each container:
- Image (required)
- Hostname
- Timeout
- environment variables
- enabling/disabling networking
- capabilities
- entry point

**Comment by rod-lin** [comment]
I wrote a scrappy deploy script here https://github.com/rod-lin/broadway-man. Should work on Ubuntu machines.

Basic usage:
`python3 man.py install master <[user@]host>` to install broadway-api on \<host\>(this command will generate and print cluster token)
`python3 man.py install worker <master host> <master port> <cluster token> <[user@]host>` to install broadway-grader on \<host\>

`python3 man.py stop master <[user@]host>` to stop master on \<host\>
`python3 man.py stop worker <[user@]host>` to stop worker on \<host\>

**Review by nmagerko — CHANGES_REQUESTED** [review]
Note that `db-url` and `port` should be specified in the `config.py`, not in `api.py`.

**Comment by rod-lin** [comment]
Hmm but I need to be able to change the parameters from command-line preferably. Passing files to containers seems messy.

**Comment by nmagerko** [comment]
Are you not going to be mounting the files in this directory anyways? I don't think I understand why this can't go in the config.

**Comment by rod-lin** [comment]
Ah I see... I'll put them the config then.

**Comment by ayushr2** [comment]
Can we close this?

**Comment by zhengyao-lin** [comment]
Outdated

**Pull request closed without merging**